Python3

reinhardt opened the pull request:
Refs #5

reinhardt committed:
Python 3: Tests passing in python 2 and python 3.
Refs #5

reinhardt committed:
Python 3: Test python 3 with travis.
Refs #5

reinhardt committed:
Replaced deprecated buildout.plonetest travis config.

ale-rt commented on src/slc/mailrouter/tests/test_injection.py:
```diff
@@ -3,7 +3,7 @@
 from zope.component.hooks import getSiteManager
 from zope.publisher.http import HTTPResponse
-from StringIO import StringIO
+from six import StringIO
```
this should be `from io import StringIO`

reinhardt replied:
Why?

ale-rt replied:
The appropriate class to store text is io.StringIO, while the appropriate class to store bytes is io.BytesIO.
Down I see that the code does:
```
mail = safe_unicode(mail)
StringIO(mail)
```
So I think that everything in this method should be textish.

reinhardt replied:
AFAICT six.StringIO does exactly the same. The only point I see is that importing from io does not rely on a third party package.

ale-rt commented on src/slc/mailrouter/storage.py:
```diff
@@ -44,7 +44,7 @@ def lookup(self, uid, _marker=None):
 
     def __getitem__(self, key):
-        return self._forward.items()[key]
+        return list(self._forward.items())[key]
```
This looks wrong, probably this was done by some 2to3 program, but it seems to me this should just be:
```
self._forward.get(key)
```

reinhardt replied:
I suspect that it's not that simple, because if it were then it could already have been done this way in python 2. But I'll look into it.

reinhardt replied:
On closer inspection it looks like this method was already broken and is never used.

reinhardt replied:
Ah, I found the spot where it is used after all. Will provide an actual fix.

reinhardt replied:
Hm. There are two spots in the code that treat the storage as a sequence of tuples. So, it does work like that, and there's not really a straightforward way to make it saner. Maybe I'll come up with something, but I won't give it highest priority.

ale-rt replied:
Oh I thought this one was changed. It seems to me we either remove it or cover with a test because I cannot see how this could work.

ale-rt commented on src/slc/mailrouter/browser/views.py:
```diff
@@ -2,7 +2,7 @@
 from tempfile import NamedTemporaryFile
 import traceback
-from StringIO import StringIO
+from six import StringIO
```
This should be from io import BytesIO

ale-rt replied:
The appropriate class to store text is io.StringIO, while the appropriate class to store bytes is io.BytesIO.
I might be tricked because I saw in the code:
```
self.request.stdin.read().encode("utf-8")
```
so I thought we were dealing with bytes.
Anyway if six.StringIO can be replaced by a more specific io class it is better

ale-rt reviewed (commented):
Added some comments here and there.
Be careful when using six.StringIO: it might not be what you expect.
Also instead of manually encoding and decoding please use safe_text, safe_nativestring and so on really clarify what is the idea there.

reinhardt committed:
Python 3: Text handling fixes.

ale-rt reviewed (changes requested):
Actually the getitem issue should be solved

reinhardt committed:
Fixed __getitem__ of storage.

reinhardt committed:
Black

reinhardt committed:
Sorted imports; re-ran black.

reinhardt committed:
Reverted __getitem__ "fix" for storage.
We actually need this convoluted line to be able to pass the storage to
the Batch class, or we need to change the call in the view.

reinhardt committed:
Python 3: Use safe_encode to be compatible with Plone 4.

ale-rt reviewed (approved):
Look goods to me but in Plone 4:
```
from Products.CMFPlone.utils import safe_encode
```
fails.
What about just dropping support for Plone 4?

reinhardt committed:
Fixed file creation.

reinhardt committed:
Removed safe_encode to be compatible with Plone 4.
We don't actually need safe_encode because we know exactly what data
type we're getting. Only one case distinction in a test is necessary.

reinhardt commented:
We don't really need safe_encode because we know what kind of data we have. I've removed/replaced it.

ale-rt commented on src/slc/mailrouter/tests/test_injection.py:
```diff
@@ -45,7 +44,9 @@ def do_inject(self, encoding=None):
         mailfile.close()
-        mail = safe_encode(mail, encoding or "utf-8")
+        if six.PY2:
+            mail = mail.decode("utf-8")
```
This will fail, you cannot write unicode in to a BytesIO:
```
[ale@emily ~]$ python2.7
Python 2.7.17rc1 (default, Oct 10 2019, 10:26:01) 
[GCC 9.2.1 20191008] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from io import BytesIO
>>> BytesIO(u"a")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'unicode' does not have the buffer interface
```

reinhardt replied:
I'm not writing it to the BytesIO like that. Look at the next line, there it's encoded again. That's to make sure we get the correct encoding that's passed to the method.

ale-rt replied:
Right! I missed it

ale-rt commented:
I see there is an `class name: adapts(IFolderish)`.
If I am not wrong that should be moved to the decorator syntax:
```
@adapter(IFolderish)
class Name:
```

reinhardt committed:
Define safe_nativestring if it can't be imported (Plone 2).

reinhardt commented:
> I see there is an `class name: adapts(IFolderish)`.

Indeed. Shouldn't that have come up in the tests or at least in the instance?

reinhardt commented:
Plone 4 is still failing. I'm giving up.

reinhardt committed:
Use adapter decorator.

reinhardt committed:
Dropped support for Plone 4.

reinhardt commented:
That travis failure is bogus (temporary error downloading buildout config). Are you happy to merge @ale-rt ?

ale-rt commented on setup.py:
```diff
@@ -1,43 +1,39 @@
-    namespace_packages=['slc'],
+        "Framework :: Plone :: 5.0",
+        "Framework :: Plone :: 5.1",
+        "Framework :: Plone :: 5.2",
```
I do not think we can claim support for anything other than `"Framework :: Plone :: 5.2"`

reinhardt replied:
Fair enough. Changed it.

ale-rt commented on src/slc/mailrouter/browser/views.py:
```diff
@@ -1,120 +1,137 @@
+
+try:
+    from Products.CMFPlone.utils import safe_nativestring
+except ImportError:
```
If we are going to support only 5.2 this BBB code can go

reinhardt replied:
True! Removing.

ale-rt reviewed (approved):
For me it is fine to merge, it will be an improvement to master.
I still have some doubts but we can resolve them later

reinhardt committed:
Let's only claim support for Plone 5.2.

reinhardt committed:
Revert "Define safe_nativestring if it can't be imported (Plone 2)."

This reverts commit 859ea5c1bc8bec2310eb36070ad225052e1c76a7.
